### PR TITLE
HDDS-5360. DN failed to process all delete block commands in one heartbeat interval

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-5360

### CODE REVIEW

**Review by jojochuang — COMMENTED** [review]
So basically make delete block operation asynchronous, moving it to another thread.
From a very quick look at the PR, looks reasonable to me.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java [hunk 1]
```diff
@@ -115,56 +201,71 @@ public void handle(SCMCommand command, OzoneContainer container,
...
-                markBlocksForDeletionSchemaV2(containerData, entry,
-                    newDeletionBlocks, entry.getTxID());
-              } else {
-                throw new UnsupportedOperationException(
-                    "Only schema version 1 and schema version 2 are "
-                        + "supported.");
+      List<Future> futures = new ArrayList<>();
+      for (int i = 0; i < containerBlocks.size(); i++) {
+        DeletedBlocksTransaction tx = containerBlocks.get(i);
+        Future future = executor.submit(() -> {
```
Comment (jojochuang) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java:207]: this is hard to read. Can you move out this Callable?

**Comment by ChenSammi** [comment]
图片

This is the rocksDB cache metrics data captured from one DN.  It costs about 200ms for open a rocksdb to persist the block deletion trasaction command.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java (hunk 1 above)
Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java]: Sure.  I refactored the code.

**Comment by jojochuang** [comment]
> 图片
> 
> This is the rocksDB cache metrics data captured from one DN. It costs about 200ms for open a rocksdb to persist the block deletion trasaction command. Becased on current default configuration, one DN can handle at most 60s / 200ms = 300 rocksdb opens if cache missed in 60s which is SCM block deleting command generation interval.

This is yet another reason why we should reconsider the 1-db-per-container design :)

**Comment by ChenSammi** [comment]
> container

Do you mean 1-db-per-volume design?  We think the same way.

**Review by jojochuang — COMMENTED** [review]
just one comment. The rest looks good to me.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java [hunk 2]
```diff
@@ -76,34 +84,178 @@
...
+            processCmd(cmd.getCmd(), cmd.getContainer(), cmd.getContext(),
+                cmd.getConnectionManager());
+          } catch (Throwable e) {
+            LOG.error("taskProcess failed.", e);
+          }
+        }
+
+        try {
+          Thread.sleep(2000);
+        } catch (InterruptedException e) {
```
Comment (jojochuang) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java:176]: Oh one more thing. We should do
`Thread.currentThread().interrupt();`

**Comment by ChenSammi** [comment]
Hi @adoroszlai ,  do we need to do something to free disk space if "Warning:  ForkStarter IOException: java.io.IOException: No space left on device" is found?

**Review by jojochuang — APPROVED** [review]
+1 from me. Waiting for Attila's review.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java [hunk 3]
```diff
@@ -76,34 +84,179 @@
...
+   */
+  public final class DeleteCmdWorker implements Runnable {
+
+    @Override
+    public void run() {
+      while (true) {
+        while (!deleteCommandQueues.isEmpty()) {
+          DeleteCmdInfo cmd = deleteCommandQueues.poll();
+          try {
+            processCmd(cmd.getCmd(), cmd.getContainer(), cmd.getContext(),
```
Comment (JacksonYao287) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java:167]: thanks for the work ! NIT, can we use `DeleteCmdInfo` as the only one parameter of `processCmd`

Comment (ChenSammi) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/DeleteBlocksCommandHandler.java:167]: Sure.

**Comment by ChenSammi** [comment]
Thanks @jojochuang and @JacksonYao287  for the code review.  Also thanks @adoroszlai for taking care of the "no space left" issue.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/TestContainerReplication.java [hunk 4]
```diff
@@ -134,7 +134,6 @@ public void testContainerReplication() throws Exception {
 
     cluster.shutdownHddsDatanode(keyLocation.getPipeline().getFirstNode());
 
-    waitForReplicaCount(containerID, 2, cluster);
```
Comment (adoroszlai) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/TestContainerReplication.java:137]: Why was this removed?  Without this, the test may give false negative: replicate count may be 3 both initially and after re-replication, and we need to distinguish between the two.

**Comment by adoroszlai** [comment]
`TestBlockDeletion` timed out twice on `master` since this commit.

* https://github.com/elek/ozone-build-results/blob/master/2021/08/10/9545/it-ozone/hadoop-ozone/integration-test/org.apache.hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion.txt
* https://github.com/elek/ozone-build-results/blob/master/2021/08/10/9555/it-ozone/hadoop-ozone/integration-test/org.apache.hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion.txt

I think this should be fixed.

**Comment by ChenSammi** [comment]
> 
> 
> `TestBlockDeletion` timed out twice on `master` since this commit.
> 
>     * https://github.com/elek/ozone-build-results/blob/master/2021/08/10/9545/it-ozone/hadoop-ozone/integration-test/org.apache.hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion.txt
> 
>     * https://github.com/elek/ozone-build-results/blob/master/2021/08/10/9555/it-ozone/hadoop-ozone/integration-test/org.apache.hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion.txt
> 
> 
> I think this should be fixed.

Sure.  I will take a look of timeout issue.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/TestContainerReplication.java (hunk 4 above)
Comment (ChenSammi) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/container/TestContainerReplication.java:137]: Because delete thread pool shutdown takes time， when it finishes， the replication manager already have the container replicated. So the wait for 2 replicas randomly fails because of there are already 3 replicas.

**Comment by adoroszlai** [comment]
> Sure. I will take a look of timeout issue.

Thanks.  Filed HDDS-5605 for it.  HDDS-5606 is also related, but it started happening earlier (few weeks ago).
